src/usr/local/www/services_dhcp.php

```diff
@@ -235,6 +242,54 @@ function validate_partial_mac_list($maclist) {
 
 	/* input validation */
 
+	/*
+	 * Check the OMAPI settings
+	 * - Make sure that if the port is defined, that it is valid and isn't in use
+	 * - Make sure the key length is at least 256bit (45 characters)
+	 * - Generate a new key if selected
+	 */
+	if (!empty($_POST['omapi_port'])) {
+		// Check the port entry
+		if (is_port($_POST['omapi_port']) && $_POST['omapi_port'] > 1024) {
+			// Get the list of open or listening ports locally
+			exec("/usr/bin/netstat -an4p tcp | /usr/bin/awk '{print $4}' | /usr/bin/egrep -o '*([0-9]{1,5})$' | /usr/bin/sort -u", $port_info);
+
+			// Check to see if the port is in the list
+			if (in_array($_POST['omapi_port'], $port_info) && $_POST['omapi_port'] != $dhcpdconf['omapi_port']){
+				$input_errors[] = gettext("Specified port number for OMAPI is in use. Please choose another port or consider using the default.");
+			}
+		} else {
+			$input_errors[] = gettext("The specified OMAPI port number is invalid. Port number must be between 1024 and 65635.");
+		}
+
+		// Generate a key if checked
+		if ($_POST['omapi_gen_key'] == "yes") {
+			// Set some vars
+			$key_bit_len = 256;
+			$key_bytes_len = $key_bit_len / 8; // 8 bits = 1 Byte
+
+			// Generate random bytes based on key length
+			$ran_bytes = openssl_random_pseudo_bytes($key_bytes_len);
+
+			// Encode the bytes to get the key string
+			$key_str = base64_encode($ran_bytes);
+
+			// Set the key
+			$_POST['omapi_key'] = $key_str;
+			$pconfig['omapi_key'] = $key_str;
+
+			// Uncheck the generate box
+			unset($_POST['omapi_gen_key']);
+			unset($pconfig['omapi_gen_key']);
+		} elseif (!empty($_POST['omapi_key'])) { // Check the key if it's not being generated
+			if (strlen($_POST['omapi_key']) < 44) {
+				$input_errors[] = gettext("Please specify a valid OMAPI key. Key must use HMAC-MD5 and be a minimum length of 256 bits.");
```

Comment:
If the server supports various HMAC types up to HMAC-SHA512, why limit the user input to HMAC-MD5? Only because that's what was hardcoded above? Consider adding a GUI option letting the user select the key algorithm as well, which would be more accommodating.

Reply:
Yea, this was due to it being hardcoded in services. I have added another input that will be used to dynamically set the alg based on user selection